daml-assistant/daml-project-config/DAML/Project/Types.hs

```diff
@@ -44,27 +46,32 @@ newtype ProjectConfig = ProjectConfig
     } deriving (Eq, Show, Y.FromJSON)
 
 newtype SdkVersion = SdkVersion
-    { unwrapSdkVersion :: Text
-    } deriving (Eq, Show, Y.FromJSON)
-
-newtype SdkChannel = SdkChannel
-    { unwrapSdkChannel :: Text
-    } deriving (Eq, Show, Y.FromJSON)
-
-newtype SdkSubVersion = SdkSubVersion
-    { unwrapSdkSubVersion :: Text
-    } deriving (Eq, Show, Y.FromJSON)
-
-splitVersion :: SdkVersion -> (SdkChannel, SdkSubVersion)
-splitVersion (SdkVersion v) =
-    case T.splitOn "-" v of
-        [] -> error "Logic error: empty version name."
-        (ch : rest) -> (SdkChannel ch, SdkSubVersion (T.intercalate "-" rest))
-
-joinVersion :: (SdkChannel, SdkSubVersion) -> SdkVersion
-joinVersion (SdkChannel a, SdkSubVersion "") = SdkVersion a
-joinVersion (SdkChannel a, SdkSubVersion b) = SdkVersion (a <> "-" <> b)
-
+    { unwrapSdkVersion :: V.Version
+    } deriving (Eq, Ord, Show)
+
+instance Y.FromJSON SdkVersion where
+    parseJSON y = do
+        verE <- V.fromText <$> Y.parseJSON y
+        case verE of
+            Left e -> fail ("Invalid SDK version: " <> e)
+            Right v -> pure (SdkVersion v)
+
+data SdkChannel
+    = Stable
+    | Unstable
+    | Custom [V.Identifier]
+    deriving (Eq, Ord, Show)
+
+versionChannel :: SdkVersion -> SdkChannel
+versionChannel (SdkVersion v) =
+    let ch = v L.^. V.release in
+    fromMaybe (Custom ch) $ lookup (map Just ch)
```

Comment:
```suggestion
    case ch of
        [] -> Stable
        [x] | Just x == V.textual "unstable" -> Unstable
        _ -> Custom ch
```

Your choice but I find the logic here quite hard to parse.

Reply:
this is much nicer, thank you :-)